### PR TITLE
feat: [CDS-70246]: Refactor decryption logic for instance Sync v2 NG

### PR DESCRIPTION
**Pull request by namantalaycha10**
## Describe your changes
With the following change we are moving the decryption logic to general class so that we do not have to decrypt for every deploymentDetails

## Local test results
1. for new deployment creating one replica
Screenshot 2023-05-25 at 6 45 50 PM

2. Manually scaling down the instance to 0
Screenshot 2023-05-26 at 10 17 21 AM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48296)

### CODE REVIEW

**Comment by namantalaycha10** [comment]
trigger uts

**Comment by namantalaycha10** [comment]
trigger uts

**Comment by namantalaycha10** [comment]
trigger uts

**Comment by namantalaycha10** [comment]
trigger messagecheck

File: 260-delegate/src/main/java/io/harness/perpetualtask/AbstractInstanceSyncV2TaskExecutor.java [hunk 1]
```diff
@@ -118,17 +128,19 @@ public PerpetualTaskResponse runOnce(
           .build();
     }
   }
-  private List<ServerInstanceInfo> retrieveServiceInstances(PerpetualTaskId taskId, PerpetualTaskExecutionParams params,
-      DeploymentReleaseDetails deploymentReleaseDetails, String accountId, InstanceSyncData.Builder instanceSyncData) {
+  private List<ServerInstanceInfo> retrieveServiceInstances(PerpetualTaskId taskId,
```
Comment (abosii-harness) [inline 260-delegate/src/main/java/io/harness/perpetualtask/AbstractInstanceSyncV2TaskExecutor.java:131]: nit: add a space before

Comment (abosii-harness) [reply 260-delegate/src/main/java/io/harness/perpetualtask/AbstractInstanceSyncV2TaskExecutor.java:131]: And can we rename this method? It does match with the abstract one, may be confusing

File: 260-delegate/src/main/java/io/harness/perpetualtask/AbstractInstanceSyncV2TaskExecutor.java [hunk 2]
```diff
@@ -171,15 +183,27 @@ private void createBatchAndPublish(AtomicInteger batchInstanceCount, AtomicInteg
...
   }
 
   protected abstract String getAccountId(@NotNull PerpetualTaskExecutionParams params);
-  protected abstract String getDeploymentType(ServerInstanceInfo serverInstanceInfos);
+
+  protected abstract InstanceSyncV2Request createRequest(String perpetualTaskId, PerpetualTaskExecutionParams params);
+
   protected abstract List<ServerInstanceInfo> retrieveServiceInstances(
-      PerpetualTaskId taskId, PerpetualTaskExecutionParams params, DeploymentReleaseDetails details);
+      PerpetualTaskId taskId, InstanceSyncV2Request instanceSyncV2Request, DeploymentReleaseDetails details);
```
Comment (abosii-harness) [inline 260-delegate/src/main/java/io/harness/perpetualtask/AbstractInstanceSyncV2TaskExecutor.java:206]: Do we need to pass taskId here? Should be already available in InstanceSyncV2Request

**Review by abosii-harness — CHANGES_REQUESTED** [review]

File: 260-delegate/src/main/java/io/harness/perpetualtask/AbstractInstanceSyncV2TaskExecutor.java (hunk 1 above)
Comment (namantalaycha10) [reply 260-delegate/src/main/java/io/harness/perpetualtask/AbstractInstanceSyncV2TaskExecutor.java:131]: thanks for pointing out, doing the changes

File: 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java [hunk 3]
```diff
@@ -104,6 +103,7 @@ public class InstanceSyncServiceImpl implements InstanceSyncService {
   private InstanceSyncServiceUtils utils;
   private InstanceSyncMonitoringService instanceSyncMonitoringService;
   private AccountClient accountClient;
+  private KryoSerializer kryoSerializer;
```
[the remark this answers is not in the document]
Comment (namantalaycha10) [reply 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java:106]: okay

**Comment by namantalaycha10** [comment]
trigger uts

**Comment by namantalaycha10** [comment]
trigger uts

**Comment by namantalaycha10** [comment]
trigger uts

**Review by abosii-harness — DISMISSED** [review]

File: 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jennyjames21) [reply 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java:106]: just to be in sync, lets use kryoSerializer for now. i can change to referenceFalse

**Review by GoelPratyush99 — DISMISSED** [review]

**Comment by namantalaycha10** [comment]
trigger uts

**Comment by namantalaycha10** [comment]
trigger review

File: 260-delegate/src/main/java/io/harness/perpetualtask/AbstractInstanceSyncV2TaskExecutor.java [hunk 4]
```diff
@@ -93,8 +103,8 @@ public PerpetualTaskResponse runOnce(
 
         for (DeploymentReleaseDetails deploymentReleaseDetails : deploymentReleaseDetailsList.getContent()) {
           InstanceSyncData.Builder instanceSyncData = InstanceSyncData.newBuilder();
-          List<ServerInstanceInfo> serverInstanceInfos =
-              retrieveServiceInstances(taskId, params, deploymentReleaseDetails, accountId, instanceSyncData);
+          List<ServerInstanceInfo> serverInstanceInfos = getServiceInstancesFromCluster(
+              instanceSyncV2Request, deploymentReleaseDetails, accountId, instanceSyncData);
```
Comment (abhinavsingh1196) [inline 260-delegate/src/main/java/io/harness/perpetualtask/AbstractInstanceSyncV2TaskExecutor.java:107]: instanceSyncV2Request already has an attribute `accountId`. Can we use that to reduce this method's parameters further?

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]

**Review by abhinavsingh1196 — APPROVED** [review]

**Review by GoelPratyush99 — APPROVED** [review]

**Pull request merged**